File: frontend/src/app/features/garden/plant.service.ts
import { Injectable, inject } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable } from 'rxjs';

export interface PlantData {
    id: string;
    nickname: string;
    species: string;
    imageUrl: string | null;
    status: 'SEED' | 'GERMINATED' | 'VEGETATIVE' | 'FLOWERING' | 'FRUITING' | 'HARVESTED' | 'DEAD' | 'ALIVE';
    ownerId: string;
    ownerName: string;
    plantedDate: string;
    harvestDate?: string;
    isVerified: boolean;
    createdAt: string;
}

export interface PlantLog {
    id: string;
    imageUrl: string | null;
    notes: string;
    logDate: string;
}

@Injectable({
    providedIn: 'root'
})
export class PlantService {
    private http = inject(HttpClient);
    private baseUrl = 'http://localhost:8080/api/v1/plants';

    addPlant(nickname: string, species: string, status: string, plantedDate: string, isVerified: boolean, image?: File): Observable<PlantData> {
        const formData = new FormData();
        formData.append('nickname', nickname);
        if (species) formData.append('species', species);
        if (status) formData.append('status', status);
        if (plantedDate) formData.append('plantedDate', plantedDate);
        formData.append('isVerified', String(isVerified));
        if (image) formData.append('image', image);
        return this.http.post<PlantData>(this.baseUrl, formData);
    }

    getUserPlants(userId: string): Observable<PlantData[]> {
        return this.http.get<PlantData[]>(`${this.baseUrl}/user/${userId}`);
    }

    getPlant(plantId: string): Observable<PlantData> {
        return this.http.get<PlantData>(`${this.baseUrl}/${plantId}`);
    }
    updatePlant(id: string, nickname: string, species: string, status: string, plantedDate: string, harvestDate?: string, image?: File): Observable<PlantData> {
        const formData = new FormData();
        formData.append('nickname', nickname);
        if (species) formData.append('species', species);
        if (status) formData.append('status', status);
        if (plantedDate) formData.append('plantedDate', plantedDate);
        if (harvestDate) formData.append('harvestDate', harvestDate);
        if (image) formData.append('image', image);

        return this.http.put<PlantData>(`${this.baseUrl}/${id}`, formData);
    }

    deletePlant(id: string): Observable<void> {
        return this.http.delete<void>(`${this.baseUrl}/${id}`);
    }

    addLog(plantId: string, notes: string, logDate: string, image?: File): Observable<PlantLog> {
        const formData = new FormData();
        if (notes) formData.append('notes', notes);
        if (logDate) formData.append('logDate', logDate);
        if (image) formData.append('image', image);
        return this.http.post<PlantLog>(`${this.baseUrl}/${plantId}/logs`, formData);
    }

    getLogs(plantId: string): Observable<PlantLog[]> {
        return this.http.get<PlantLog[]>(`${this.baseUrl}/${plantId}/logs`);
    }

    updateCoverPhoto(plantId: string, image: File): Observable<PlantData> {
        const formData = new FormData();
        formData.append('image', image);
        return this.http.put<PlantData>(`${this.baseUrl}/${plantId}/photo`, formData);
    }

    deleteLog(logId: string): Observable<void> {
        return this.http.delete<void>(`${this.baseUrl}/logs/${logId}`);
    }

    updateLog(logId: string, notes: string): Observable<PlantLog> {
        return this.http.put<PlantLog>(`${this.baseUrl}/logs/${logId}`, { notes });
    }
}

File: backend/src/main/java/com/plantsocial/backend/service/PlantService.java
package com.plantsocial.backend.service;

import com.plantsocial.backend.dto.PlantResponse;
import com.plantsocial.backend.model.Plant;
import com.plantsocial.backend.model.PlantStatus;
import com.plantsocial.backend.repository.PlantRepository;
import com.plantsocial.backend.user.User;
import com.plantsocial.backend.user.UserRepository;
import com.plantsocial.backend.model.PlantLog;
import com.plantsocial.backend.repository.PlantLogRepository;
import com.plantsocial.backend.dto.LogResponse;
import lombok.RequiredArgsConstructor;
import org.springframework.security.core.context.SecurityContextHolder;
import org.springframework.security.core.userdetails.UserDetails;
import org.springframework.security.core.userdetails.UsernameNotFoundException;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;
import org.springframework.web.multipart.MultipartFile;

import java.util.List;
import java.util.UUID;
import java.util.stream.Collectors;

@Service
@RequiredArgsConstructor
public class PlantService {

    private final PlantRepository plantRepository;
    private final UserRepository userRepository;
    private final com.plantsocial.backend.repository.PostRepository postRepository;
    private final PlantLogRepository plantLogRepository;
    private final FileStorageService fileStorageService;

    @Transactional
    public PlantResponse addPlant(String nickname, String species, String status, java.time.LocalDate plantedDate,
            boolean isVerified, MultipartFile image) {
        User user = getCurrentUser();

        String imageUrl = null;
        if (image != null && !image.isEmpty()) {
            imageUrl = fileStorageService.storeFile(image);
        }

        PlantStatus plantStatus = PlantStatus.VEGETATIVE;
        if (status != null && !status.isBlank()) {
            try {
                plantStatus = PlantStatus.valueOf(status.toUpperCase());
            } catch (IllegalArgumentException e) {
                // Keep default
            }
        }

        java.time.LocalDate pDate = (plantedDate != null) ? plantedDate : java.time.LocalDate.now();

        Plant plant = Plant.builder()
                .owner(user)
                .nickname(nickname)
                .species(species)
                .imageUrl(imageUrl)
                .status(plantStatus)
                .plantedDate(pDate)
                .isVerified(isVerified)
                .build();
        Plant saved = plantRepository.save(plant);
        return mapToResponse(saved);
    }

    public List<PlantResponse> getUserPlants(UUID userId) {
        return plantRepository.findByOwnerIdOrderByCreatedAtDesc(userId)
                .stream()
                .map(this::mapToResponse)
                .collect(Collectors.toList());
    }

    public PlantResponse getPlant(UUID plantId) {
        Plant plant = plantRepository.findById(plantId)
                .orElseThrow(() -> new IllegalArgumentException("Plant not found"));
        return mapToResponse(plant);
    }

    @Transactional
    public PlantResponse updatePlant(UUID plantId, String nickname, String species, String status,
            java.time.LocalDate plantedDate, java.time.LocalDate harvestDate, MultipartFile image) {
        Plant plant = plantRepository.findById(plantId)
                .orElseThrow(() -> new IllegalArgumentException("Plant not found"));

        if (nickname != null && !nickname.isBlank()) {
            plant.setNickname(nickname);
        }
        if (species != null) {
            plant.setSpecies(species);
        }
        if (status != null && !status.isBlank()) {
            try {
                plant.setStatus(PlantStatus.valueOf(status.toUpperCase()));
            } catch (IllegalArgumentException e) {
                // Ignore invalid status
            }
        }
        if (plantedDate != null) {
            plant.setPlantedDate(plantedDate);
        }
        if (harvestDate != null) {
            plant.setHarvestDate(harvestDate);
        }

        if (image != null && !image.isEmpty()) {
            String imageUrl = fileStorageService.storeFile(image);
            plant.setImageUrl(imageUrl);
        }

        Plant saved = plantRepository.save(plant);
        return mapToResponse(saved);
    }

    @Transactional
    public void deletePlant(UUID plantId) {
        // 1. Unlink posts
        List<com.plantsocial.backend.model.Post> posts = postRepository.findAllByPlantId(plantId);
        for (com.plantsocial.backend.model.Post post : posts) {
            post.setPlant(null);
        }
        postRepository.saveAll(posts);

        // 2. Delete plant logs
        List<PlantLog> logs = plantLogRepository.findAllByPlantIdOrderByLogDateDesc(plantId);
        plantLogRepository.deleteAll(logs);

        // 3. Delete plant
        plantRepository.deleteById(plantId);
    }

    private PlantResponse mapToResponse(Plant plant) {
        return new PlantResponse(
                plant.getId(),
                plant.getNickname(),
                plant.getSpecies(),
                plant.getImageUrl(),
                plant.getStatus().name(),
                plant.getOwner().getId(),
                plant.getOwner().getFullName(),
                plant.getPlantedDate(),
                plant.getHarvestDate(),
                plant.isVerified(),
                plant.getCreatedAt());
    }

    @Transactional
    public com.plantsocial.backend.dto.LogResponse addLog(UUID plantId, String notes, java.time.LocalDate logDate,
            MultipartFile image) {
        Plant plant = plantRepository.findById(plantId)
                .orElseThrow(() -> new IllegalArgumentException("Plant not found"));

        String imageUrl = null;
        if (image != null && !image.isEmpty()) {
            imageUrl = fileStorageService.storeFile(image);
        }

        PlantLog log = PlantLog.builder()
                .plant(plant)
                .notes(notes)
                .logDate(logDate != null ? logDate : java.time.LocalDate.now())
                .imageUrl(imageUrl)
                .build();

        PlantLog saved = plantLogRepository.save(log);
        return new com.plantsocial.backend.dto.LogResponse(saved.getId(), saved.getImageUrl(), saved.getNotes(),
                saved.getLogDate());
    }

    public List<com.plantsocial.backend.dto.LogResponse> getPlantLogs(UUID plantId) {
        return plantLogRepository.findAllByPlantIdOrderByLogDateDesc(plantId)
                .stream()
                .map(log -> new com.plantsocial.backend.dto.LogResponse(log.getId(), log.getImageUrl(), log.getNotes(),
                        log.getLogDate()))
                .collect(Collectors.toList());
    }

    @Transactional
    public PlantResponse updatePlantPhoto(UUID plantId, MultipartFile image) {
        Plant plant = plantRepository.findById(plantId)
                .orElseThrow(() -> new IllegalArgumentException("Plant not found"));

        User currentUser = getCurrentUser();
        if (!plant.getOwner().getId().equals(currentUser.getId())) {
            throw new org.springframework.security.access.AccessDeniedException("Not authorized");
        }

        if (image != null && !image.isEmpty()) {
            String imageUrl = fileStorageService.storeFile(image);
            plant.setImageUrl(imageUrl);
        }

        Plant saved = plantRepository.save(plant);
        return mapToResponse(saved);
    }

    @Transactional
    public void deleteLog(UUID logId) {
        PlantLog log = plantLogRepository.findById(logId)
                .orElseThrow(() -> new IllegalArgumentException("Log not found"));

        User currentUser = getCurrentUser();
        if (!log.getPlant().getOwner().getId().equals(currentUser.getId())) {
            throw new org.springframework.security.access.AccessDeniedException("Not authorized");
        }
        plantLogRepository.delete(log);
    }

    @Transactional
    public com.plantsocial.backend.dto.LogResponse updateLog(UUID logId, String notes) {
        PlantLog log = plantLogRepository.findById(logId)
                .orElseThrow(() -> new IllegalArgumentException("Log not found"));

        User currentUser = getCurrentUser();
        if (!log.getPlant().getOwner().getId().equals(currentUser.getId())) {
            throw new org.springframework.security.access.AccessDeniedException("Not authorized");
        }

        if (notes != null) {
            log.setNotes(notes);
        }

        PlantLog saved = plantLogRepository.save(log);
        return new com.plantsocial.backend.dto.LogResponse(saved.getId(), saved.getImageUrl(), saved.getNotes(),
                saved.getLogDate());
    }

    private User getCurrentUser() {
        Object principal = SecurityContextHolder.getContext().getAuthentication().getPrincipal();
        String username;
        if (principal instanceof UserDetails) {
            username = ((UserDetails) principal).getUsername();
        } else {
            username = principal.toString();
        }
        return userRepository.findByUsername(username)
                .or(() -> userRepository.findByEmail(username))
                .orElseThrow(() -> new UsernameNotFoundException("User not found"));
    }
}
